### ISSUE TITLE
Kubernetes controller ignores label selector when selecting gateway classes

### ISSUE THREAD

**Issue by WillDaSilva** [traefik/traefik#11155]
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.traefik.io) and didn't find any.

### What did you do?

I am using Traefik with the Kubernetes Gateway API.

Per [the GatewayClass docs](https://gateway-api.sigs.k8s.io/api-types/gatewayclass/) I am using two GatewayClass resources in my Kubernetes cluster: one for internal services, and one for internet-facing services.

To deploy this, I deployed the Traefik Helm chart twice into my cluster: one named `traefik-internal` into the `traefik-internal` namespace, and one named `traefik-external`into the `traefik-external`. The configuration used is included below.

Notably the following configuration is set:
- For the `traefik-internal` release:
  - `providers.kubernetesGateway.labelselector: app.kubernetes.io/instance=traefik-internal-traefik-internal`
  - `gatewayClass.name: traefik-internal`
  - and additional arguments for the controller:
    - "--providers.kubernetesgateway.statusaddress.service.namespace=traefik-internal"
    - "--providers.kubernetesgateway.statusaddress.service.name=traefik-internal-internal"
- For the `traefik-external` release:
  - `providers.kubernetesGateway.labelselector: app.kubernetes.io/instance=traefik-external-traefik-external`
  - `gatewayClass.name: traefik-external`
  - and additional arguments for the controller:
    - "--providers.kubernetesgateway.statusaddress.service.namespace=traefik-external"
    - "--providers.kubernetesgateway.statusaddress.service.name=traefik-external-external"

I expected the label selector to cause each Traefik controller to only use the GatewayClass with the matching label. I expected the status address to be set appropriately for each gateway, to the hostname known to each controller respectively.

### What did you see instead?

Both of my Gateway resources have `status.addresses[0].value` set to the hostname of the `traefik-internal` load balancer, or both are set to the hostname of the `traefik-external` load balancer.

I believe the label selector I have configured (to make the `traefik-internal` controller use the `traefik-internal` GatewayClass, and to make the `traefik-external` controller use the `traefik-external` GatewayClass) is being ignored. This is resulting in a race condition where both Traefik controllers attempt to accept both gateways. Which ever one gets to both of them first will set both of their status address to the address it has.

https://github.com/traefik/traefik/blob/a7502c8700e111963b2b52fb0d25c0128fb0c05a/pkg/provider/kubernetes/gateway/client.go#L382

Looking at the above code, it seems like the label selector is ignored. This function is used by this code:

https://github.com/traefik/traefik/blob/a7502c8700e111963b2b52fb0d25c0128fb0c05a/pkg/provider/kubernetes/gateway/kubernetes.go#L315-L345

In this code the controller iterates over each GatewayClass, and accept it so long as the controller name matches a particular hardcoded value. The controller sets the message "Handled by Traefik controller", though unfortunately it does not specify which Traefik controller is handling the GatewayClass. If that information were included, then this issue would be easier to debug. As it stands, it merely seems like both GatewayClass resources are being accepted by the same controller, rather than one each. 

### What version of Traefik are you using?

`v3.1.4` (via chart version `32.0.0`)

### What is your environment & configuration?

`helmfile.yaml`:

```yaml
helmDefaults:
  timeout: 600
  wait: true

repositories:
  - name: traefik
    url: https://traefik.github.io/charts
  - name: eks
    url: https://aws.github.io/eks-charts

releases:
  - name: aws-load-balancer-controller
    namespace: kube-system
    createNamespace: false
    chart: eks/aws-load-balancer-controller
    version: "1.8.3"
    values:
      - clusterName: "<cluster name>"
        serviceAccount:
          create: true
          name: aws-load-balancer-controller
          annotations:
            eks.amazonaws.com/role-arn: "arn:aws:iam::<account ID>:role/<controller role name>"
        region: "<region>"
        vpcId: "<VPC ID>"
    hooks:
      - events: ["prepare"]
        command: "kubectl"
        args:
          [
            "apply",
            "--server-side",
            "--force-conflicts",
            "-k",
            "https://github.com/aws/eks-charts/stable/aws-load-balancer-controller/crds",
          ]
  - name: traefik-external # Traefik instance to expose services to the public internet
    namespace: traefik-external
    createNamespace: true
    chart: traefik/traefik
    version: "32.0.0"
    hooks:
      - events: ["prepare"]
        command: "kubectl"
        args:
          [
            "apply",
            "--server-side",
            "--force-conflicts",
            "-k",
            "https://github.com/traefik/traefik-helm-chart/traefik/crds/",
          ]
    values:
      - providers:
          kubernetesCRD:
            enabled: false
          kubernetesIngress:
            enabled: false
          kubernetesGateway:
            enabled: true
            labelselector: "app.kubernetes.io/instance=traefik-external-traefik-external"
        gatewayClass:
          name: traefik-external
        additionalArguments:
          - "--providers.kubernetesgateway.statusaddress.service.namespace=traefik-external"
          - "--providers.kubernetesgateway.statusaddress.service.name=traefik-external-external"
        service:
          enabled: false # disable the default service
          additionalServices:
            external:
              type: LoadBalancer
              annotations:
                service.beta.kubernetes.io/aws-load-balancer-scheme: internet-facing
                service.beta.kubernetes.io/aws-load-balancer-backend-protocol: tcp
                service.beta.kubernetes.io/aws-load-balancer-type: nlb
                service.beta.kubernetes.io/aws-load-balancer-nlb-target-type: ip
                service.beta.kubernetes.io/aws-load-balancer-ip-address-type: dualstack
                service.beta.kubernetes.io/aws-load-balancer-cross-zone-load-balancing-enabled: "true"
        ports:
          web: null
          websecure: null
          http-external:
            expose:
              external: true
              internal: false
            port: 8000
            exposedPort: 80
            protocol: TCP
          https-external:
            expose:
              external: true
              internal: false
            port: 8443
            exposedPort: 443
            protocol: TCP
        gateway:
          name: traefik-external-gateway
          listeners:
            web: null
            # NOTE: The websecure listener is disabled by default; don't set it to null,
            #       because if you do that'll create a new key, rather than delete an
            #       existing one.
            http-external:
              port: 8000
              protocol: HTTP
              namespacePolicy: All
            https-external:
              port: 8443
              protocol: HTTPS
              namespacePolicy: All
              mode: Terminate
              certificateRefs:
                - kind: Secret
                  name: base-wildcard-tls-cert
                  namespace: cert-manager
        logs:
          general:
            level: DEBUG
            format: json # "common", "json", or null
          access:
            enabled: true
            format: json # "common", "json", or null
            addInternals: true # Log access for Traefik's internal services (e.g. the dashboard)
  - name: traefik-internal # Traefik instance to expose services on the VPN
    namespace: traefik-internal
    createNamespace: true
    chart: traefik/traefik
    version: "32.0.0"
    hooks:
      - events: ["prepare"]
        command: "kubectl"
        args:
          [
            "apply",
            "--server-side",
            "--force-conflicts",
            "-k",
            "https://github.com/traefik/traefik-helm-chart/traefik/crds/",
          ]
    values:
      - providers:
          kubernetesCRD:
            enabled: false
          kubernetesIngress:
            enabled: false
          kubernetesGateway:
            enabled: true
            labelselector: "app.kubernetes.io/instance=traefik-internal-traefik-internal"
        gatewayClass:
          name: traefik-internal
        additionalArguments:
          - "--providers.kubernetesgateway.statusaddress.service.namespace=traefik-internal"
          - "--providers.kubernetesgateway.statusaddress.service.name=traefik-internal-internal"
        service:
          enabled: false # disable the default service
          additionalServices:
            internal:
              type: LoadBalancer
              annotations:
                service.beta.kubernetes.io/aws-load-balancer-scheme: internal
                service.beta.kubernetes.io/aws-load-balancer-backend-protocol: tcp
                service.beta.kubernetes.io/aws-load-balancer-type: nlb
                service.beta.kubernetes.io/aws-load-balancer-nlb-target-type: ip
                service.beta.kubernetes.io/aws-load-balancer-ip-address-type: dualstack
                service.beta.kubernetes.io/aws-load-balancer-cross-zone-load-balancing-enabled: "true"
        ports:
          web: null
          websecure: null
          http-internal:
            expose:
              external: false
              internal: true
            port: 8001
            exposedPort: 80
            protocol: TCP
          https-internal:
            expose:
              external: false
              internal: true
            port: 8444
            exposedPort: 443
            protocol: TCP
        gateway:
          name: traefik-internal-gateway
          listeners:
            web: null
            # NOTE: The websecure listener is disabled by default; don't set it to null,
            #       because if you do that'll create a new key, rather than delete an
            #       existing one.
            http-internal:
              port: 8001
              protocol: HTTP
              namespacePolicy: All
            https-internal:
              port: 8444
              protocol: HTTPS
              namespacePolicy: All
              mode: Terminate
              certificateRefs:
                - kind: Secret
                  name: base-wildcard-tls-cert
                  namespace: cert-manager
        logs:
          general:
            level: DEBUG
            format: json # "common", "json", or null
          access:
            enabled: true
            format: json # "common", "json", or null
            addInternals: true # Log access for Traefik's internal services (e.g. the dashboard)
```

This can be applied to an EKS cluster by running `helmfile apply`. This bug is not specific to EKS, but I've only attempted to reproduce this behaviour on EKS so far. In an effort to make this example more self-contained, I moved the values in-line in the Helmfile above, and replaced my actual values with angle bracket enclosed placeholders.

**Comment by WillDaSilva** [comment]
@kevinpollet Thanks for triaging this issue. Is there anything I could do to help resolve this issue?

**Comment by kevinpollet** [comment]
Hello @WillDaSilva and thanks for raising this,

We opened #11169 to fix this issue, it will be available in the next v3.2 version.
In the meantime the fix will be available in the [traefik/traefik:experimental-v3.2](https://hub.docker.com/layers/traefik/traefik/experimental-v3.2/images/sha256-4175271e44825726dabf41f1ef94cd9509627933752da930782ce85179bcb5d5?context=explore) image once the pr is merged.

**Comment by WillDaSilva** [comment]
I can confirm this is fixed in Traefik v3.2. Thank you!

Exposing the Traefik dashboard for one instance via another instance still seems broken, but that warrants another issue.

**Comment by kevinpollet** [comment]
Glad to hear it, feel free to open another issue if needed.
